src/operator/tensor/cast_storage-inl.h

```diff
@@ -120,9 +119,7 @@ struct CastStorageRspDnsKernel {
     IType rid = idx[i];
     dim_t dns_offset = rid * row_length;
     dim_t rsp_offset = i * row_length;
-    for (dim_t col = 0; col < row_length; col++) {
-      dns[dns_offset + col] = data[rsp_offset + col];
-    }
+    memcpy(dns + dns_offset, data + rsp_offset, sizeof(DType) * row_length);
```

Comment:
Thanks for the contribution. I don't think this will work on GPU though as this kernel is shared by CPU and GPU..

Reply:
Maybe do
```c++
#if defined(__CUDACC__)
    cudaMemcpyAsync(...);
#else
    memcpy(...);
#endif
```

Reply:
@szha  You cannot call `cudaMemcpyAsync` inside a cuda kernel. The only way to copy elements inside cuda kernel is to perform elemwise assignment.

Reply:
Did a bit of searching on this and the following turns up:
1. memcpy seems to work in https://stackoverflow.com/questions/10456728/is-there-an-equivalent-to-memcpy-that-works-inside-a-cuda-kernel (might be outdated)
2. cudaMemcpyAsync on device is available since CUDA6, though there are [limitations](http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#ixzz4rwhsHEfu)
```
Notes about all memcpy/memset functions:
Only async memcpy/set functions are supported
Only device-to-device memcpy is permitted
May not pass in local or shared memory pointers
```

Reply:
Thanks for your reply. I checked the `cudaMemcpyAsync`  [method](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79) in cuda 8.0 and found that now it is not limited to device-to-device memcpy.

Reply:
@yaochengji nice. APIs are getting better by the day.

Reply:
BTW do you mind sharing your reference? I'm interested.

Reply:
@szha [link](http://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79)
Please find details in its description.